toolchain/check/function.cpp

```diff
@@ -94,18 +99,23 @@ auto CheckFunctionReturnType(Context& context, SemIRLoc loc,
     return context.emitter().Build(loc, IncompleteTypeInFunctionReturnType,
                                    return_type_id);
   };
+  SemIR::Function::ReturnSlot result;
   if (!context.TryToCompleteType(
           return_type_id,
           function.return_slot == SemIR::Function::ReturnSlot::Error
               ? std::nullopt
               : std::optional(diagnose_incomplete_return_type))) {
-    function.return_slot = SemIR::Function::ReturnSlot::Error;
+    result = SemIR::Function::ReturnSlot::Error;
   } else if (SemIR::GetInitRepr(context.sem_ir(), return_type_id)
                  .has_return_slot()) {
-    function.return_slot = SemIR::Function::ReturnSlot::Present;
+    result = SemIR::Function::ReturnSlot::Present;
   } else {
-    function.return_slot = SemIR::Function::ReturnSlot::Absent;
+    result = SemIR::Function::ReturnSlot::Absent;
+  }
+  if (!specific_id.is_valid()) {
+    function.return_slot = result;
```

Comment:
Should we have return slot caching for specifics?

Reply:
I'm not sure the caching is even worthwhile for non-generic functions. I don't think we look at the return slot information for a function frequently enough for this to matter, and it's pretty cheap anyway (mostly just a hash table lookup to find if the type is complete). The functional effect of the caching is that we only diagnose an incomplete return type once, but perhaps we can do something more narrowly targeted if we think that's important.

Anyway, that's for a future PR I think :)